### ISSUE TITLE
Feature Request: option to bake fallbacks into locale files (fallbacks option in v3)

### ISSUE THREAD

**Issue by oleksii-leonov** [fnando/i18n-js#658]
## Description

@fnando, thank you for your great work 👍 

I really like `v4`. Design decisions in `v4` align with how we use `v3` in our company (export each locale to separate static JSON file, then import locale file with `webpacker`).

But I found one missing feature that complicates our migration from `v3` to `v4`.

We have substantial localization files (~1MB raw JSON for one locale, ~150KB compressed) and ~20 locales.

Several major locales are updated in sync; other locales are pulled from 3-rd party localization service and are usually a bit behind major locales. So, we need to use fallbacks.

In `v3` we have fallbacks baked in localization JSON files. And we need to load only one localization file (~150KB compressed) for the user.

`v4` export localization files without fallbacks baking. So, we need to load two locales and then use the JS-side fallback feature. It feels like a huge overhead to load two localization files (150KB + 150KB) just to get a fallback for a dozen of strings.

## Describe the solution

It would be nice to have the option to bake fallbacks into locale files. The same as `fallbacks` option in `v3`.
I could try to do a PR for such an option if it's aligned with the `v4` vision.

## Alternatives you considered

It's possible to merge JSON locale files after export with a custom JS or Ruby script.
But it's not trivial `deep_merge`. Domain-specific logic is required: https://github.com/fnando/i18n-js/blob/4b07b30e8b6f42c4415f4bb967e8e2986a0e2268/lib/i18n/js/utils.rb#L44

Merging locales on the Ruby side feels easier and more straightforward.
Also, I believe this issue is relevant for many library users, and it's nice to have it in the library.

As another option, I mind about adding the hook `before_export`.
This hook would be very useful for modifying the locale before exporting to JSON (now it's possible only with monkey-patch).

**Comment by fnando** [comment]
I think we should implement some sort of plugin api that can be used to extend i18n-js without having to necessarily tweak the main package.

I'm thinking of adding something like this:

```ruby
class FallbackPlugin < I18nJS::Plugin
  def self.validate_schema(config)
    # called while validating configuration schema.
    # plugin can perform config validations, like checking for required keys related.
  end

  def self.setup
    # called by I18nJS when activating the plugin. This can be used to set up config requirements.
    I18nJS::Schema.root_keys << :fallback
    I18nJS::Schema.required_root_keys << :fallback
  end

  def self.transform(translations)
    # return the transformed translations
    translations
  end
end

I18nJS.register_plugin(FallbackPlugin)
```

I'm going to play with this idea and see where it goes!

**Comment by fnando** [comment]
@oleksii-leonov I've implemented an initial version of the Plugin API and a embed fallback translation plugin on #676. Can you please test it and see if it's working as expected?

**Comment by oleksii-leonov** [comment]
@fnando , thank you, it looks great 👍 
I will try to test it this week.

**Comment by oleksii-leonov** [comment]
@fnando 

It works, but for large locales, it becomes too slow.
I have extracted `deep_merge` from v3, and now it works instantly — please check #688.

**Comment by fnando** [comment]
Just released #688 via v4.2.1.
